Update Action Groups API

ritwik8119 opened the pull request:
# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#sdk-generation-guidelines)
- [x] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [x] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [x] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/main/eng/mgmt/mgmtmetadata).
- [x] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

fengzhou-msft committed:
Merge from upstream main

Commit:
Update Action Groups API

Commit:
Updates for SDK generation and management metadata

Commit:
Generated code for Monitor namespace and updated script

ritwik8119 commented on sdk/monitor/Microsoft.Azure.Management.Monitor/src/Generated/Operations.cs:
```diff
@@ -84,7 +84,7 @@ public async Task<AzureOperationResponse<OperationListResult>> ListWithHttpMessa
             }
             // Construct URL
             var _baseUrl = Client.BaseUri.AbsoluteUri;
-            var _url = new System.Uri(new System.Uri(_baseUrl + (_baseUrl.EndsWith("/") ? "" : "/")), "providers/microsoft.insights/operations").ToString();
```
@HarveyLink After updating the generate script, it pulled in changes for multiple other APIs like this one, with changes to the URI. The issue is that these changes are now breaking tests and causing builds to fail. Example test failure in the build (https://dev.azure.com/azure-sdk/public/_build/results?buildId=1169052&view=logs&j=ca395085-040a-526b-2ce8-bdc85f692774&t=3374e909-b3a0-5fde-4fba-1f1bb546637f):

 System.Collections.Generic.KeyNotFoundException : Unable to find a matching HTTP request for URL 'GET /providers/Microsoft.Insights/operations?api-version=2015-04-01'. Calling method Item().
                       Stack Trace:
                          at Microsoft.Azure.Test.HttpRecorder.Records.get_Item(String key)
                        at Microsoft.Azure.Test.HttpRecorder.HttpMockServer.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)


Could you please advise on how to fix these tests?

Commit:
Fix failing tests

HarveyLink commented:
@ritwik8119 The proper fix here would be re-record the failing tests and commit the record file. Or, if it is truely case-sensitive issue, then you may try replace all the related test records with correct one.

ritwik8119 commented:
> @ritwik8119 The proper fix here would be re-record the failing tests and commit the record file. Or, if it is truely case-sensitive issue, then you may try replace all the related test records with correct one.

@HarveyLink I tried setting the correct values in the test records in this commit, but it appears that the tests are still failing with the same error: https://github.com/Azure/azure-sdk-for-net/pull/24855/commits/1f00f80dd2d10ddd9b1cc06f2e9177a2c36b5912

Was this the right place to update the records, or is there any other location to update?
If this was the right place, then it appears directly updating the records does not fix the failure. Is there any documentation for how re-record these tests? (as I am not familiar with the process of recording them)

archerzz commented:
@ritwik8119 You can refer to the test framework doc： https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/core/Azure.Core.TestFramework

Commit:
Fix failing tests

ritwik8119 commented:
@archerzz @HarveyLink I have fixed the failing tests. However, the build is again failing due to discrepancy in the code. I added the "AutoRestCodeGenerationFlags" flag in the generate.ps1 script as you mentioned (https://github.com/Azure/azure-sdk-for-net/pull/24855/files#diff-9e71eba224e3746cd458ad43314129f8897e318f8083fe5799ac8ca0d847ea5b):
![image](https://user-images.githubusercontent.com/21337621/139774134-a29728e8-6c8c-436e-9f1f-8faad7b6f68d.png)

However it appears that the build is not picking up this flag as we see this in the build logs (https://dev.azure.com/azure-sdk/public/_build/results?buildId=1172370&view=logs&j=ca395085-040a-526b-2ce8-bdc85f692774&t=50b9345f-2a7f-5980-cb0f-1a8f9fb09dd0): 

For reference, we are using following commands for the code check: 
autorest https://github.com/Azure/azure-rest-api-specs/blob/abf7b7e10780d737d99ef0013274431d2024a17d/specification/monitor/resource-manager/readme.md --csharp --version=v2 --reflect-api-versions --csharp-sdks-folder=D:/a/1/s/sdk  --use:@microsoft.azure/autorest.csharp@2.3.82
 
Could you please advice on how to fix this?

HarveyLink commented:
@ritwik8119 For the metadata issue, please make sure you are using the latset `Start-AutoRestCodeGeneration` command. Pull latset main, and then re-install the module using: `dotnet msbuild build.proj /t:Util /p:UtilityName=InstallPsModules`

ritwik8119 commented:
> @ritwik8119 For the metadata issue, please make sure you are using the latset `Start-AutoRestCodeGeneration` command. Pull latset main, and then re-install the module using: `dotnet msbuild build.proj /t:Util /p:UtilityName=InstallPsModules`

Sorry, but which metadata issue do you mean? The discrepancy here appears to caused by the reference command not including the tags which I passed in generate.ps1: 
![image](https://user-images.githubusercontent.com/21337621/139791400-958df66f-4065-4178-96a4-23833442ff37.png)

this should include the "--tag=package-2021-09" which I have passed, but it doesnt. Could you please help figure out why the reference command is not including this tag?

HarveyLink commented:
@ritwik8119 The command we use for comparing generate code in pipeline, is from the metadata file logged by the generation tool. In this case it is: [autorest https://github.com/Azure/azure-rest-api-specs/blob/abf7b7e10780d737d99ef0013274431d2024a17d/specification/monitor/resource-manager/readme.md --csharp --version=v2 --reflect-api-versions --csharp-sdks-folder=D:/a/1/s/sdk  --use:@microsoft.azure/autorest.csharp@2.3.82](https://dev.azure.com/azure-sdk/public/_build/results?buildId=1172370&view=logs&j=ca395085-040a-526b-2ce8-bdc85f692774&t=50b9345f-2a7f-5980-cb0f-1a8f9fb09dd0&l=164)
We have update out generation tool to support customized tag, which need to update the `Start-AutoRestCodeGeneration` using the command I mentioned.
BTW, the metadata looks incorrect to me because it still useing `Azure` repo and `master` branch, which should be `main` by default. Please make sure you get the latest command and inlcude the change while committing your code. In your case it should be a fork repo and your own branch.
![Screenshot 2021-11-02 151534](https://user-images.githubusercontent.com/30464227/139802088-9a07b117-0522-4d49-90dd-be2ed430def5.png)
.

Commit:
Revert "Merge from upstream main"

This reverts commit 00f54e85c5c2b6df47696e79e7e02ed4b1a61702.

Commit:
Merge from upstream main

Commit:
Revert "Merge from upstream main"

This reverts commit 252c7ff67aef9229bb7fe32b0162ebaa3dba62a5.

Commit:
Run latest Start-AutoRestCodeGeneration

ritwik8119 commented:
> @ritwik8119 The command we use for comparing generate code in pipeline, is from the metadata file logged by the generation tool. In this case it is: [autorest https://github.com/Azure/azure-rest-api-specs/blob/abf7b7e10780d737d99ef0013274431d2024a17d/specification/monitor/resource-manager/readme.md --csharp --version=v2 --reflect-api-versions --csharp-sdks-folder=D:/a/1/s/sdk --use:@microsoft.azure/autorest.csharp@2.3.82](https://dev.azure.com/azure-sdk/public/_build/results?buildId=1172370&view=logs&j=ca395085-040a-526b-2ce8-bdc85f692774&t=50b9345f-2a7f-5980-cb0f-1a8f9fb09dd0&l=164) We have update out generation tool to support customized tag, which need to update the `Start-AutoRestCodeGeneration` using the command I mentioned. BTW, the metadata looks incorrect to me because it still using `Azure` repo and `master` branch, which should be `main` by default. Please make sure you get the latest command and inlcude the change while committing your code. In your case it should be a fork repo and your own branch. ![Screenshot 2021-11-02 151534](https://user-images.githubusercontent.com/30464227/139802088-9a07b117-0522-4d49-90dd-be2ed430def5.png) .

@HarveyLink Thank you for the update, I ran with the latest version and the code verify has passed. However there are some checks that seem to have been canceled because of timeout errors in the pipeline tasks, which I believe should not be related to the code change, like this one: https://dev.azure.com/azure-sdk/public/_build/results?buildId=1174582&view=logs&j=3dc8fd7e-4368-5a92-293e-d53cefc8c4b3
![image](https://user-images.githubusercontent.com/21337621/139959908-03e32af6-d5a9-4b56-9a21-b9d9ce78e469.png)

https://dev.azure.com/azure-sdk/public/_build/results?buildId=1174581&view=logs&j=e44bd064-d8d9-5feb-6343-09b8c49ab1a8
![image](https://user-images.githubusercontent.com/21337621/139960032-82c1a471-22c6-4e3e-b857-78993a5b24f6.png)

Could you please help fix this?

ritwik8119 commented on sdk/monitor/Microsoft.Azure.Management.Monitor/src/Microsoft.Azure.Management.Monitor.csproj:
```diff
@@ -6,13 +6,11 @@
   <PropertyGroup>
     <Description>Microsoft Azure Monitor Library</Description>
     <AssemblyName>Microsoft.Azure.Management.Monitor</AssemblyName>
-    <Version>0.25.4-preview</Version>
```
@HarveyLink @archerzz While I added this change as it was earlier requested in this review, I see that there have been multiple breaking changes in the past which did not bump up the major version to 1. Would it suffice if I bumped up the version to 0.26.0 (instead of 1.0.0)?

HarveyLink replied:
Since it is a sider update rather than a major api version change, I agreed with your minor version 0.26.0-preview. @archerzz What's your opinion?

archerzz replied:
@ritwik8119  I'm fine. Also if your next version is still `-preview`, then you're free to use `0.26.0-preview` since it's not official release. But you need mention this to the team for their awareness that we should avoid breaking changes after GA. Thanks.

ritwik8119 commented:
/check-enforcer evaluate

ritwik8119 commented:
/azp run

ritwik8119 commented:
> @ritwik8119 The command we use for comparing generate code in pipeline, is from the metadata file logged by the generation tool. In this case it is: [autorest https://github.com/Azure/azure-rest-api-specs/blob/abf7b7e10780d737d99ef0013274431d2024a17d/specification/monitor/resource-manager/readme.md --csharp --version=v2 --reflect-api-versions --csharp-sdks-folder=D:/a/1/s/sdk --use:@microsoft.azure/autorest.csharp@2.3.82](https://dev.azure.com/azure-sdk/public/_build/results?buildId=1172370&view=logs&j=ca395085-040a-526b-2ce8-bdc85f692774&t=50b9345f-2a7f-5980-cb0f-1a8f9fb09dd0&l=164) We have update out generation tool to support customized tag, which need to update the `Start-AutoRestCodeGeneration` using the command I mentioned. BTW, the metadata looks incorrect to me because it still using `Azure` repo and `master` branch, which should be `main` by default. Please make sure you get the latest command and inlcude the change while committing your code. In your case it should be a fork repo and your own branch. ![Screenshot 2021-11-02 151534](https://user-images.githubusercontent.com/30464227/139802088-9a07b117-0522-4d49-90dd-be2ed430def5.png) .

@HarveyLink Regarding using Azure-main instead of my fork, I realized that there were multiple changes to other APIs that were not checked into the SDK repo by the time I had created my fork already, and using my fork would make no difference. Moreover, I chose the specific tag where all other APIs were at their last stable versions, and only updated my API, so the only updates pulled in with the auto-generate were for minor changes like the camel-casing of request URIs. Running the latest Start-AutoRestCodeGeneration has fixed the verification errors and all the checks are passing now. Could you please re-review?

HarveyLink reviewed (commented):
LGTM

archerzz commented:
@ritwik8119 LGTM. You can update the version, then merge. Thanks.

Commit:
Update new version
